src/Microsoft.NetCore.Analyzers/Core/Tasks/DoNotCreateTaskCompletionSourceWithWrongArguments.cs

```diff
@@ -0,0 +1,89 @@
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessage,
+                                                                             DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: null,
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // Ensure necessary types are available
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskCompletionSource, out var tcsType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemObject, out var objectType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTaskContinuationOptions, out var taskContinutationOptionsType))
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    // Ensure that you're using the TaskCompletionSource constructor with the object state property else exit
+                    var invocation = (IObjectCreationOperation)operationContext.Operation;
+                    if (!invocation.Type.OriginalDefinition.Equals(tcsType) ||
+                        invocation.Arguments.Length == 0 ||
+                        !invocation.Constructor.Parameters[0].Type.Equals(objectType))
+                    {
+                        return;
+                    }
+
+                    // Check to see if argument is a conversion operator
+                    if (!(invocation.Arguments[0] is IArgumentOperation arg) ||
+                        arg.Value.Kind != OperationKind.Conversion ||
+                        !(arg.Value is IConversionOperation conversionOperation))
```

Comment:
@mavasani, what's the recommendation for checking Kind before doing a cast?  This is mixing both: for invocation.Arguments[0], it's just doing the type check, but then for arg.Value it's doing both the kind check and then the cast check.

Reply:
For performance sensitive code, it is recommended to do a Kind check followed by an explicit direct cast to the interface. This is mandated for all code written inside the Roslyn compiler code base. However, for majority of the analyzers, we also allow any of the below alternate code styles:

1. `is` expression - does not require kind check
2. `as` expression followed by a null check

Reply:
So anything with the above 4 lines of code need to change?